src/dwt.rs

```diff
@@ -0,0 +1,207 @@
+    pub fn measure<F: FnOnce()>(&self, f: F) -> ClockDiff {
+        let mut times: [u32; 2] = [0; 2];
+        let mut sw = self.stopwatch(&mut times);
+        f();
+        sw.lap().last()
+    }
+    pub fn release(self) -> (DWT, DCB) {
+        (self.dwt, self.dcb)
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct Delay {
+    dwt_clock: Hertz,
+}
+impl Delay {
+    pub fn delay_cycles(cycles: u64) {
+        Delay::delay_cycles_(DWT::get_cycle_count(), cycles);
+    }
+    fn delay_us_(self, us: u64) {
+        let start = DWT::get_cycle_count();
+        let cycles = (us * self.dwt_clock.0 as u64) / 1_000_000;
+        Delay::delay_cycles_(start, cycles);
+    }
+    fn delay_ms_(self, ms: u64) {
+        let start = DWT::get_cycle_count();
+        let cycles = (ms * self.dwt_clock.0 as u64) / 1_000;
+        Delay::delay_cycles_(start, cycles);
+    }
+    fn delay_cycles_(mut start: u32, cycles: u64) {
+        if cycles < (u32::MAX / 2) as u64 {
+            let cycles = cycles as u32;
+            while (DWT::get_cycle_count().wrapping_sub(start)) < cycles {}
+        } else if cycles <= u32::MAX as u64 {
+            let mut cycles = cycles as u32;
+            cycles -= u32::MAX / 2;
+            while (DWT::get_cycle_count().wrapping_sub(start)) < u32::MAX / 2 {}
+            start -= u32::MAX / 2;
+            while (DWT::get_cycle_count().wrapping_sub(start)) < cycles {}
+        } else {
+            let mut rest = (cycles >> 32) as u32;
+            let cycles = (cycles & u32::MAX as u64) as u32;
+            loop {
+                while (DWT::get_cycle_count().wrapping_sub(start)) < cycles {}
+                if rest == 0 {
+                    break;
+                }
+                rest -= 1;
+                while (DWT::get_cycle_count().wrapping_sub(start)) > cycles {}
+            }
+        }
+    }
+}
+
+// Implement DelayUs/DelayMs for various integer types
+macro_rules! impl_DelayIntT {
+    (for $($t:ty),+) => {$(
+        impl DelayUs<$t> for Delay {
+            fn delay_us(&mut self, us: $t) {
+                self.delay_us_(us as u64);
+            }
+        }
+        impl DelayMs<$t> for Delay {
+            fn delay_ms(&mut self, ms: $t) {
+                self.delay_ms_(ms as u64);
+            }
+        }
+    )*}
+}
+impl_DelayIntT!(for usize,u64,u32,u16,u8,i64,i32,i16,i8);
+
+/// Implement DelayS for float types (Note, that the delay itself will need to calculate a lot)
+macro_rules! impl_DelayFloatT {
+    (for $($t:ty),+) => {$(
+        impl DelayS<$t> for Delay {
+            fn delay(&mut self, s:$t) {
+                let start = DWT::get_cycle_count();
+                let cycles = s * self.dwt_clock.0 as $t;
+                Delay::delay_cycles_(start, cycles as u64);
+            }
+        }
+    )*}
+}
+impl_DelayFloatT!(for f32,f64);
+
+/// Very simple stopwatch
+/// Notes:
+/// - Max tdiff is < (1<<32) / HCLK
+/// - Assumes DWT cycle counter is enabled
+/// - Assumes Debug is set up
```

Comment:
Is that correct and what does that mean?

Reply:
- The 32-bit cycle counter will overflow very soon
(eg. after ~25s@168MHz)
- Instanciating StopWatch does not enable
    - the cycle counter (`DWT.enable_cycle_counter()`)
    - nor the trace unit (`DCB.enable_trace()`)
    - it just assumes the `dcb.constrain(..)` stuff was done before